Notes/Algorithms/Data Structures/Hash-Tables.c

```diff
@@ -1,230 +1,244 @@
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 #include <stdbool.h>
 
 #define SUCCESS 0
 #define ERROR -1
-#define OVERFLOW -2
-#define EMPTY -3
-#define TEST 9
-#define EXPECTED 10
-#define SIZE 10
 
-// Hash struct
+// Hash Item struct
 typedef struct Item {
-    int key;
+    char* key;
     int value;
 } Item;
 
-//create global testHash and expectedHash for testFunctions to use
-Item* testHash[SIZE];
-Item* expectedHash[SIZE];
+// Hash Table struct
+typedef struct HashTable {
+    int size;
+    Item** items;
+} HashTable;
 
-void traverse() {
-    for (int i = 0; i < SIZE; i++) {
-        int value = testHash[i]->value;
-        int key = testHash[i]->key;        
-        printf("Test Key: %d Value: %d \n", key, value);
-    }
-    for (int i = 0; i < SIZE; i++) {
-        int value = expectedHash[i]->value;
-        int key = expectedHash[i]->key;
-        printf("Expected Key: %d Value: %d \n", key, value);
+// returns true if a collision would happen for the initially hashed index for a key/value pair and false if no collision would occurred 
+bool isCollision(HashTable * table, int index) {
+    if (table->items[index] != NULL) {
+        return true;
     }
+    return false;
 }
 
-//compare checks the Items
-bool compare() {
-    for (int i = 0; i < SIZE; i++) {
-        int testKey = testHash[i]->key;
-        int testValue = testHash[i]->value;
-        int expectedKey = expectedHash[i]->key;
-        int expectedValue = expectedHash[i]->value;
+// compare checks the testHash and expectedHash for parity
+// if the two HashTables match returns true
+// if a mismatch occurs returns false
+bool compare(HashTable * testHash, HashTable * expectedHash) {
+    for (int i = 0; i < 10; i++) {
+        int testKey = testHash->items[i]->key;
+        int testValue = testHash->items[i]->value;
+        int expectedKey = expectedHash->items[i]->key;
+        int expectedValue = expectedHash->items[i]->value;
         if (testKey != expectedKey || testValue != expectedValue) {
             return false;
         }
     }
     return true;
 }
 
-int hashing(int key) {
-    return key % SIZE;
+// hashing 
+// returns a "hashed" index based on modulus of the key's ASCII integer sum and the SIZE of the hashTable
```

Comment:
I'm glad you're exploring the territory of what it means to hash a string. ASCII int sum (a.k.a. "loselose") seems like an interesting choice... but I'm not convinced it has good key distribution.  Why this over something like djb2?  I expect choices like this to have some "why" backup in the comments.  e.g:

```// I chose int sum for simplicity and clarity since this is a demo.  I'd use X in the wild```

Some brief discussion on the topic as well as other concise implementations with better distribution here:

http://www.cse.yorku.ca/~oz/hash.html

Reply:
Per our conversation - I used loselose because it was a simple hashing function I should on StackOverflow and it seemed adequate for the purposes of this. If I had seen the handy link provided I would have utilized djb2 instead. I made a comment in the code explaining why this choice was made! [Commit f7e2d2fb821d1deedb858f8539c0ba3b7991f528](https://github.com/kcarrel/Training-Notes/pull/12/commits/f7e2d2fb821d1deedb858f8539c0ba3b7991f528)